File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline

t = int(input())

for _ in range(t):
    ps = list(input().rstrip())
    check = []
    cnt = True
    for i in ps:
        if i == "(":
            check.append(1)
        else:
            if len(check) == 0:
                cnt = False
                break
            else:
                check.pop()
                
    if len(check) > 0:
        cnt = False

    if cnt == True:
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws Exception {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		
		for (int i = 0; i < n; i++) {
			Stack<String> stack = new Stack<>();
			boolean isTrue = true;
			String str = br.readLine();
			for (int j = 0; j < str.length(); j++) {
				if (str.charAt(j) == '(') {
					stack.add("(");
				} else {
					if (stack.isEmpty()) {
						isTrue = false;
						break;
					} else {
						stack.pop();
					}
				}
			}
			
			if (!stack.isEmpty()) isTrue = false;
			if (isTrue) System.out.println("YES");
			else System.out.println("NO");
		}

	}

}
